### ISSUE TITLE
Stack traces can be interleaved when running multithreaded

### ISSUE THREAD

**Issue by kpamnany** [JuliaLang/julia#51056]
If exceptions occur in tasks running on 2 (or more) threads and we try to output backtraces concurrently, the output can have interleaved lines, making it difficult to figure out the actual stack for either exception.

Probably the best (only?) solution is to replace the use of [`jl_safe_printf()`](https://github.com/JuliaLang/julia/blob/master/src/stackwalk.c#L622) wherever it is used, and instead accumulate the entire backtrace into a buffer, then use a single `jl_safe_printf` to output the buffer.

Any other solutions?

**Comment by NHDaly** [comment]
Another solution (and the patch we're going with for now in our own builds, just to get _something_ going here) is to prepend the thread id, as in `thread (4)` before every stacktrace line. This way the user can disentangle them in post-processing, or just with their eyes.

I can see either one being sufficient. I would be a bit afraid of buffering, in case one of the threads hits some other fatal error and kills the whole process before we managed to output any of the buffers? 😬

**Comment by gbaraldi** [comment]
I agree with Kiran, prepending the thread isn't the way to go. The distributed tests do that by printing worker before and it's unreadable. Maybe there is a way to make it a bit safer, but you are always at risk of dying in the middle of printing anyway.

**Comment by vtjnash** [comment]
printing the Task id (the address of the Task object) seems like a good idea to me
